Tools/pyco/pyco.py

```diff
@@ -306,6 +306,9 @@ def rewritten_bytecode(code: types.CodeType, builder: Builder) -> bytes:
                             assert False, \
                                 f"{value} is not an immediately loadable constant"
             else:
+                if is_function_code(code):
+                    assert oparg > 0
+                    oparg -= 1
```

Comment:
this is why the functions were crashing.

Reply:
Ignore me, this is not the right fix.

Reply:
The problem with this is that we could have an opcode with oparg 0 if, say, the docstring is None and there is another access to None in the function. So I tried to remove the thing in CodeObject.preload() that excludes code.co_consts[0] from the constants for functions (so we always include the docstring).  But then it totally corrupt functions and I haven't figured out why.  I'll continue tomorrow.

Reply:
> (The handler is the case block in the switch in ceval.c.)

In this case the handler invokes the subroutine and tells it to proceed to the next opcode. So we don't really return from the subroutines. Maybe we need to change that.

Reply:
> In this case the handler invokes the subroutine and tells it to proceed to the next opcode. So we don't really return from the subroutines. Maybe we need to change that.

For this I think I have a solution. In the case for LAZY_LOAD_CONSTANT, first try
```
value = GETITEM(consts, oparg);
```
If that's not NULL, we have the value (INCREF, PUSH, DISPATCH). Otherwise, try what's currently there. **If that's not NULL,** INCREF and set the value in consts (i.e. co->co_consts), and then INCREF, PUSH, DISPATCH.

Finally, call the subroutine with the existing code. When the subroutine returns, it sets the value in pyc->consts and pushes it onto the stack, but leaves co->co_consts unchanged. That's where the added check and code kicks in -- **the next time we hit the same instruction,** it will copy the value from pyc->consts to co->co_consts. From then on we'll take the first (fast) path whenever we hit this particular instruction.

IOW:
```diff
diff --git a/Python/ceval.c b/Python/ceval.c
index 916fcf1f2e..3ecc0152f8 100644
--- a/Python/ceval.c
+++ b/Python/ceval.c
@@ -4365,12 +4365,20 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
         }

         case TARGET(LAZY_LOAD_CONSTANT): {
-            PyObject *value = GETITEM(co->co_pyc->consts, co->co_consts_start + oparg); 
+            PyObject *value = GETITEM(consts, oparg);
             if (value != NULL) {
                 Py_INCREF(value);
                 PUSH(value);
                 DISPATCH();
             }
+            value = GETITEM(co->co_pyc->consts, co->co_consts_start + oparg);
+            if (value != NULL) {
+                Py_INCREF(value);
+                PyTuple_SET_ITEM(consts, oparg, value);
+                Py_INCREF(value);
+                PUSH(value);
+                DISPATCH();
+            }

             _Py_CODEUNIT *instrs;
             int stacksize;
```

This does not solve the big problem, I'm still stumped by that but have some more time.

Reply:
Maybe we could just work with co->co_pyc->consts and populate co->co_consts when it is requested. It's only accessed in a handlful of places from the C code, so we could add a C function to populate and return it.